Code to address Issues - #9, #12, #13

takposha opened the pull request:
This PR primarily handles topic extraction as outlined in #9.  BERTopic is primarily used alongside KeyBERT and LangChain to extract topics discussed in a video transcript over time, and the corresponding model for topic extraction as well. This will be used for generating questions in a future Issue. 
It also fixes issues raised in #12 and #13, 
For #12: A separate Python script  `captionsProcessor.py` has been added that performs all the steps similar to the `captionsProcessor.ipynb` notebook. As of now, variables for configuration are modified in the script itself. Maybe a future issue can be raised for adding more configurability from environmental variables. 
For #13: The `helper.py` script has been split into smaller modular files to better reflect the functions' uses for different parts of the code. These include the `transcriptLoader.py`, `topicExtractor.py`, and `utils.py`.

takposha committed:
fixes #9, #13

- Implemented Topic Extraction from transcripts using BERTopic
- Created new .py files for saving data, topic extraction, transcript loading, and config classes
- Add new dependencies to requirements.txt
- Updated ReadMe file with new steps for Topic Extraction

takposha committed:
Added python script to address Issue #12

Added printAndLog function to utils.py for easier logging of runtime messages

Replaced logging functions with printAndLog calls in topicExtractor.py

takposha committed:
MInor fixes to RwadMe text and messaging in topicExtractor.

takposha committed:
Expanded ReadMe to cover info on the seperate Python scripts created.

pushyamig commented on requirements.txt:
```diff
@@ -1,3 +1,12 @@
 python-dotenv==1.0.1
-IPython==8.12.3
+scikit_learn==1.4.2
+tiktoken==0.6.0
```
There is some issue with `tiktoken` package version. 

```
ERROR: Cannot install -r requirements.txt (line 6) and tiktoken==0.6.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested tiktoken==0.6.0
    langchain-openai 0.1.7 depends on tiktoken<1 and >=0.7

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```

pushyamig replied:
is there a reason you chose that particular version. Can we just use that is being used by `langchain-openai`?

takposha replied:
No reason, this was just what the versions were on my machine when I extracted them via `pip freeze`.

pushyamig replied:
Remove it the exclusive pkg declaration and just import since it is already available package in the venv

pushyamig replied:
I think I will resolve it

pushyamig commented on captionsProcessor.py:
```diff
@@ -0,0 +1,25 @@
+    config = Config()
+    config.setFromEnv()
+
+    videoToUse = "New Google Assignments in Canvas"
```
I would suggest this value as configurable value rather having it in code. You could make this a list, if there is a plan to run it multiple video caption at time.

takposha replied:
You are right, I mentioned this in the ReadMe as well, and never got to implementing it. 
I'll keep it as a single file for now. We could later add another script that does this over a list by calling this script? I think that will help to set up multiprocessing if we wanted that.

pushyamig replied:
>We could later add another script that does this over a list by calling this script?

Sure

takposha committed:
Removed exlcusive version rqmts. for tiktoken
Added .env variable for videoToUse in config
Updated ReadMe and Captionsrocessor files for change to videoToUse variable.

takposha commented:
The above commit handles the comments raised in the PR. However, some things based on this that I will address in the next PR:
1. Rework the config class to better handle the variables being used. I think I can improve how the variables are being passed between functions to be a bit clearer. 
2. Insert more robust checks for valid and existing files and authentications. There are no checks to validate if a video file exists, or if credentials are valid, which needs to be implemented.

pushyamig commented on captionsProcessor.py:
```diff
@@ -0,0 +1,25 @@
+
+    printAndLog(f"Retrieving Topics for {videoToUse}")
+
+    topicsOverTime, topicModel = retrieveTopics(videoToUse, transcriptToUse, config=config)
```
Before passing the `transcriptToUse` response to `retrieveTopics` make some basic validation check. what if `transcriptToUse` returns `None` or whatever not suited for next function call?.  This is best practice and should account for error case as well.

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+        line = line.strip()
+        if line.isdigit():
+            continue
+        elif "-->" in line:
```
Optional: make --> as constant. when ever a string is repeated it should be made a constant. So that you don't need to retype multiple times

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+        elif "-->" in line:
+            startTime, endTime = line.split("-->")
+            startTime = datetime.strptime(startTime.strip(), "%H:%M:%S,%f")  # .time()
+            endTime = datetime.strptime(endTime.strip(), "%H:%M:%S,%f")  # .time()
```
Optional: make "%H:%M:%S,%f" as constant. when ever a string is repeated it should be made a constant. So that you don't need to retype multiple times

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+            )
+            sentence = ""
+
+    return pd.DataFrame(transcript)
```
I would make this a variable and then return the variable that hold the `df`. good for debugging purposes. also log what is returned as debug

```
transcript_df = pd.DataFrame(transcript)
return transcript_df
```

pushyamig replied:
Same applies to L80

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+def getCombinedTranscripts(
+    videoNames,
+    captionsFolder="Captions",
+    windowSize=30,
```
I like the way you are using docstring, I think you could consider using Python Typing when you feel convenient that way you don't need to mention return types, for a functions. https://docs.python.org/3/library/typing.html

But I won't stress on it now since the code is build up and it might impede the progress.

pushyamig replied:
Usually the windowSize can also be configurable option.

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+                    If multiple video names are provided, returns a dictionary of combined transcripts for each video.
+    """
+
+    if type(videoNames) == str:
```
The check here should be made if the videoNames is empty string and if it is then log and exit the process. Since without the videoname it doesn't make sense to proceed further

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+
+    srtFiles, transcripts, sentences, combinedTranscripts = {}, {}, {}, {}
+    for video in videoNames:
+        srtFiles[video] = glob.glob(f"{captionsFolder}/{video}/*.srt")
```
Surround this with try/catch since the error here could be file not found if the user did not configured the file. Again, log/exit the process

takposha committed:
Fixed indentation error `config.py` file at line 33.

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+
+    Returns:
+        dict or str: If only one video name is provided, returns the combined transcript as a string.
+                    If multiple video names are provided, returns a dictionary of combined transcripts for each video.
```
I like the way you are using docstring, I think you could consider using Python Typing when you feel convenient that way you don't need to mention return types, for a functions. https://docs.python.org/3/library/typing.html

But I won't stress on it now since the code is build up and it might impede the progress.

pushyamig commented on captionsProcessor.py:
```diff
@@ -0,0 +1,25 @@
+
+    printAndLog(f"Retrieving Topics for {config.videoToUse}")
+
+    topicsOverTime, topicModel = retrieveTopics(config.videoToUse, transcriptToUse, config=config)
```
Have some Validation check around ` transcriptToUse ` value returned before passing it to `  retrieveTopics` This is essential and make the error tracking easier. As always code should be written to  take care of happy and unhappy path. The error case should give the user enough info to understand the issue and fix it.

pushyamig commented on transcriptLoader.py:
```diff
@@ -0,0 +1,116 @@
+    if len(videoNames) == 1:
+        return combinedTranscripts[videoNames[0]]
+
+    return combinedTranscripts
```
I feel the `combinedTranscripts` should always returning a dict either single video or multiple. having a consistence return type is needed. The upstream code which is consuming the  `combinedTranscripts` value should assume it return dict

takposha replied:
I agree. I am going to change this to not even handle multiple transcripts actually, as it makes it harder to set up multiprocessing, Instead, I am working on having something like a `transcriptData` class that stores all the variables and extracted data for a single transcript at a time. All returned values will be consistent as well that way.

pushyamig commented on topicExtractor.py:
```diff
@@ -0,0 +1,169 @@
+
+
+def retrieveTopics(
+    videoToUse, transcriptToUse, config, overwrite=False, saveNameAppend=""
```
If `videoToUse` as a configurable option, then `config` object is already holding this value. I don't think there is a need for passing the `VideoToUse` as parameter here. What do you say?

takposha replied:
You are right. I am making a Class file to just store all the transcript information instead so it is just one object to manage.

takposha replied:
Implemented in the new [Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/14/commits/12992ec3fadf5a4771519490c1f580d4f442db7b).

The `Config` class contains all the config data, which is used to create a `TranscriptData` class for all the extracted transcript data to be used in future steps.

pushyamig commented on topicExtractor.py:
```diff
@@ -0,0 +1,169 @@
+        is not False
+        and (
+            topicsOverTime := dataLoader(
+                config, "topicsOverTime", videoToUse, saveNameAppend
```
Can you explain what is the difference b/w `topicModel` and `topicsOverTime`?I think I am not clear
 by those means. I feel it's good idea to explain here as well.

takposha replied:
I will add an explanation in the ReadMe if that works.
topicModel is the actual Model itself, while topicsOverTime are the extracted topics from said model. 
There is a compute cost to getting the extracted topics, and for debugging and visualization, there is a need to keep the topic Model itself so we can call it for any other uses. Hence they are separate entities.

pushyamig replied:
I feel just write it in the function docstring here since this keywords are first mentioned as part of this code.

takposha replied:
I have made changes in the new [Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/14/commits/12992ec3fadf5a4771519490c1f580d4f442db7b). This puts them together into a single class, so it better reflects that they are part of the same object and connected to each other. It should make it clearer, if not, I will elaborate further in the `TopicModeller`'s docstring.

pushyamig commented on config.py:
```diff
@@ -141,6 +30,22 @@ def __init__(self):
         if not os.path.exists(self.captionsFolder):
             os.makedirs(self.captionsFolder)
```
Surround by try/catch accounting for file not found error.

takposha replied:
Similar to my comment below, 
This is to make a folder if it doesn't exist, like when the code is run for the very first time. Afterwards, it will run as expected. 
Is there a case that I missed where this could throw an error?

pushyamig replied:
I will resolve this.

pushyamig commented on config.py:
```diff
@@ -141,6 +30,22 @@ def __init__(self):
+        self.videoToUse = ''
+
+        self.saveFolder = "savedData"
+        self.fileTypes = ["topicModel", "topicsOverTime"]
```
Optional: `"topicModel", "topicsOverTime"` make these as constants and import whenever needed

takposha replied:
So make it a dictionary like:
self.fileTypes = {'Topic Model':"topicModel", "Topics Over Time":"topicsOverTime"}
So that even if the file names are changed, nothing else in the code needs to be edited?

pushyamig replied:
This is what I meant: https://stackoverflow.com/questions/6343330/importing-a-long-list-of-constants-to-a-python-file. you could create `constants.py` where all the constants can be listed and exported if used in multiple files. If a constant used in a single file that declare at the top of the file. But this is optional.

takposha replied:
Ohhh. I like this methodology, but I suppose only if it's a lot of variables. I will make one if there are a lot of constants I define.

pushyamig commented on config.py:
```diff
@@ -141,6 +30,22 @@ def __init__(self):
+        for folder in self.fileTypes:
+            folderPath = os.path.join(self.saveFolder, folder)
+            if not os.path.exists(folderPath):
+                os.makedirs(folderPath)
```
Again, some basic error checking for file/folder doesn't exist error

takposha replied:
Isn't the error checking being handled by making the folders if they do not exist? My thought here was to just make the folders if they do not exist, to avoid any trouble later on with the data loader and saver.

pushyamig replied:
Yes, you are correct. `os.path.exists(folderPath)` is the helper function that is doing that check for you Consider this resolved

pushyamig commented on config.py:
```diff
@@ -141,6 +30,22 @@ def __init__(self):
+                os.makedirs(folderPath)
+
+        self.representationModelType = "langchain"
+        self.useKeyBERT = True
```
Can you explain what is the purpose of this variable `useKeyBERT`?

takposha replied:
It's a variable that toggles the usage of the KeyBERT vectorizer in the BERTopic model. I kept is as an option if we ever want to toggle it. This will also be added to the .env file, defaulting to True.

takposha replied:
I have added this is a toggle option in the .env file to also explain what it does,

pushyamig commented on config.py:
```diff
@@ -1,124 +1,13 @@
+# from langchain_community.callbacks import get_openai_callback
 
 
 class Config:
```
I feel for `Config` class should use of `printAndLog`. I like the way you created this helper function and mananging the logging in `utils.py`.   Consistency patterns should be followed.

takposha replied:
I will implement this in more places. I do not want to flood the Notebook with too many INFO messages when not needed, but I can add a parameter to the printAndLog to ignore printing if asked for so the Notebook output stays clean, but the logger will keep logging.

takposha replied:
Implemented in the new [Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/14/commits/12992ec3fadf5a4771519490c1f580d4f442db7b).
All logging statements should have been replaced by the `printAndLog` function. Not all of them print, but they all do log additional data where relevant, like some descriptions of dataframes and their heads, for example.

pushyamig commented on topicExtractor.py:
```diff
@@ -0,0 +1,169 @@
+            )
+        )
+        is not False
+    ):
```
Optional: the block of code from L27-33 is not reading friendly. 

I would extract `dataLoader(config, "topicModel", videoToUse, saveNameAppend))` and `dataLoader(
                config, "topicsOverTime", videoToUse, saveNameAppend
            )` as separate variable and make the `if` call simple to read.

takposha replied:
Implemented in the new [Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/14/commits/12992ec3fadf5a4771519490c1f580d4f442db7b). 
I was trying the new Walrus operator in Python out and liked the way it worked. But for readability, not so much.

pushyamig replied:
>trying the new Walrus operator in Python out

I like the use of it. But I felt code readability is important. This has 3 `and` operator and I had halt and focus and see what's going in the code.

takposha replied:
I agree.
I feel like this is more of a list comprehension or string-slicing feature that lets you do a lot in one line of code but at the cost of readability. Has its uses, not everywhere though.

takposha committed:
Fixes for comments from PRs #8 & #14, and fixes for Issue #12.

Transcript loading and processing is handled in a Class object now.
Multiple checks have been put in place for checking existence of correct files and validating credentials.

takposha commented:
There is a bug in the `TopicModeller` class when it comes to saving that I haven't been able to replicate consistently and squash, but I am looking into it, 
In some cases, the BERTopic model will fail to save, with an IoError stating that files are already open. 
This appears to happen when the `TopicModeller`  object hasn't been destroyed and a new one tries to access the same model again, but I need to test this to confirm. The solution here will be to simply delete the whole object on completion of the process. This isn't done here as we need it for the future question generation component. 

But I would like to look more into this, as model saving, and tracking token counts is not as clean a solution as I'd hoped for.

takposha committed:
topicExtractor.py: Added Links to BERTopic and Count Vectorization Documentation & Explanations.

Updated `overwrite` to be a  'OVERWRITE_EXISTING_TOPICMODEL' configurable variable to toggle BERTopic models to be rerun or not.

pushyamig reviewed (approved):
This is looking Good. I was able to ran the code successfully. Thanks for adding link to addition link to TopicModel, TopicOverTime Etc.  The details given and walk through sessions were useful. 

I like that fact the you are using GAI and some Model training to reduce the hallucination that could be caused by GAI. 

Having a smaller PR next time would make the review process easier.

takposha commented:
> Having a smaller PR next time would make the review process easier.

That is the plan, hopefully. The next PR will be for just the question generation, which should be much smaller. 

I will get this merged and get that going next.
